### PR TITLE
Refine DataLoader support multi-processing

### PR DESCRIPTION
**Pull request by heavengate**
**Refine DataLoader support multi-processing**
- add `paddle.io.Dataset` base class
- add `paddle.io.BatchSampler`
- add `paddle.io.DataLoader` support multi-process when initialize by `__init__`
    - `num_workers = 0` for single-process mode
    - `num_workers > 0` for multi-process mode

**multi-process work flow**

bj-3ca43dab9357bbfe9d9b45595f4d5c8a5f485d9d

**data order keeping**
- multi-process workers keep data order by `reorder_dict` and `index` check
- `py_reader` can keep order by setting `is_ordered=True` after https://github.com/PaddlePaddle/Paddle/pull/22699 merged

**speed test**
- CUDA Driver version: 418.39
- CUDA: V9.0.176
- cuDNN: 7.5.1
- GPU: 8 * P40
- CPU: 24 Core Intel(R) Xeon(R) CPU E5-2650 v4 @ 2.20GHz

- orignal DataLoader configs:
1. static graph mode trained by ParallelExecutor, disable multiprocess implement in model readers
2. dynamic graph mode trained by `paddle.distributed.launch`,  will initial a DataLoader for each GPU and each DataLoader will start a subprocess
- multiprocess DataLoader configs:
1. static graph mode trained by ParallelExecutor, disable multiprocess implement in model readers, multiprocess DataLoader set `num_workers=16`
2. dynamic graph mode trained by `paddle.distributed.launch`, will initial a DataLoader for each GPU and set `num_workers=2` for each DataLoader

- static graph

| Model | batch_size  | original DataLoader | multiprocess DataLoader | speed up ratio |
| :-- | :--: | :--: | :--: | :--: |
| ResNet50 | 8\*32 |  711ms/step | 266ms/step | 167.3% |
| YOLOv3-DarkNet53 | 8\*8 | 3155ms/step | 708ms/step | 345.6% |
| TSM-ResNet50 | 8\*16 | 8665ms/step | 1669ms/step | 419.2% |
| BMN | 8\*16 |  2393ms/step | 772ms/step | 209.9% |

- dynamic graph

| Model | batch_size | origin DataLoader | multiprocess DataLoader | speed up ratio |
| :-- | :--: | :--: | :--: | :--: |
| ResNet50 | 8\*32 | 269ms/step | 267ms/step | - |
| YOLOv3-DarkNet53 | 8\*8 |  751ms/step | 623ms/step | 20.5% |
| TSM-ResNet50 | 8\*16 |  3550ms/step | 1767ms/step | 89.6% |
| BMN | 8\*16 |  831ms/step | 785ms/step | 5.8% |

- `ResNet50` `TSM` used `xmap_reader`， while `YOLOv3` `BMN` didn't

cn doc: https://github.com/PaddlePaddle/FluidDoc/pull/1952

**Document**
- DataLoader
dataloader1
dataloader2
dataloader3
dataloder4
dataloader5

- Dataset
dataset

- BatchSampler
batch_sampler1
batch_sampler2

### CODE REVIEW

**Review by xyzhou-puck — DISMISSED** [review]
LGTM.

File: python/paddle/fluid/dataloader/__init__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/__init__.py]: 2020

File: python/paddle/fluid/dataloader/dataloader_iter.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+
+from .. import core
+from ..framework import in_dygraph_mode
+from ..multiprocess_utils import CleanupFuncRegistrar, _cleanup_mmap, python_exit_flag
+
+# multi-process worker check indices queue interval, avoid
+# hanging in subprocess data loading
+MP_INDICES_CHECK_INTERVAL = 5
+
+__all__ = ['_DataLoaderIterSingleProcess', '_DataLoaderIterMultiProcess']
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/dataloader_iter.py]: there is no need to put _func in all?

File: python/paddle/fluid/dataloader/dataloader_iter.py [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+                      self._worker_init_fn, i))
+            worker.daemon = True
+            worker.start()
+            self._workers.append(worker)
+            self._worker_status.append(True)
+
+        core._set_process_pids(id(self), tuple(w.pid for w in self._workers))
+        self._set_worker_signal_handler(worker)
+
+    def _set_worker_signal_handler(self, worker):
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/dataloader_iter.py]: worker is not used in this function

File: python/paddle/fluid/dataloader/dataset.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        return path
+
+    if download:
+        return paddle.dataset.common.download(url, module_name, md5)
+    else:
+        raise FileNotFoundError(
+            '{} not exists and auto download disabled'.format(path))
+
+
+class MnistDataset(Dataset):
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/dataset.py]: MnistDataset -> MNIST ？  and  add usage

File: python/paddle/fluid/dataloader/dataset.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    """
+    Implement of mnist dataset
+
+    Args:
+        root(str): mnist dataset root path.
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/dataset.py]: there is no root in init func

File: python/paddle/fluid/dataloader/dataset.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    TEST_LABEL_MD5 = 'ec29112dd5afa0611ce80d1b7f02629c'
+    TRAIN_IMAGE_URL = URL_PREFIX + 'train-images-idx3-ubyte.gz'
+    TRAIN_IMAGE_MD5 = 'f68b3c2dcbeaaa9fbdd348bbdeb94873'
+    TRAIN_LABEL_URL = URL_PREFIX + 'train-labels-idx1-ubyte.gz'
+    TRAIN_LABEL_MD5 = 'd53e105ee54ea40749a09fcbcd1e9432'
+
+    def __init__(self,
+                 image_filename=None,
+                 label_filename=None,
+                 mode='train',
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/dataset.py]: 1. 不同的dataset还是单独文件mnist.py
2. 针对，train和test，觉得也可以封装2个函数：

def train():
    return MNIST(mode='train')

def test():
    return MNIST(mode='test')

使用时:

train = fluid.dataset.mnist.train()
test = fluid.dataset.mnist.test()

可以看下是否需求

File: python/paddle/fluid/tests/unittests/test_map_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+                pass
+
+            try:
+                dataloader.set_batch_generator()
+                self.assertTrue(False)
+            except:
+                pass
+
+
+# -------------- Dataset unittests --------------
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_map_dataloader.py]: dataset的单测可以单独分开，后续加cifar之类，还会加单测

Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_map_dataloader.py]: dataset的单测可以单独分开，后续加cifar之类，还会加单测

**Review by heavengate — COMMENTED** [review]
@qingqing01 Done，thanks

File: python/paddle/fluid/dataloader/dataset.py (hunk 6 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/dataset.py]: ` fluid.dataset.mnist.train`之前的mnist数据集实现占用了，放在`fluid.io.dataloader.minist`下面的话有点奇怪。。。

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/dataloader/batch_sampler.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                implement or other python object which implemented
+                `__len__` for BatchSampler to get indices as the
+                range of :attr:`data_source` length. Default None.
+        indices (list|tuple): a substitution parameter for
+                :attr:`data_source` either :attr:`data_source` or
+                :attr:`indices` should be set, give the whole
+                indices to sampler from directly. Default None.
+        shuffle(bool): whther to shuffle indices order before genrate
+                batch indices. :attr:`shuffle` should not be set if 
+                :attr:`sample_iter` is set. Default False.
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/batch_sampler.py]: 1. genrate -> generating.
2. sample_iter is not in init func, but used in  TestBatchSamplerWithIndicesAndDataSource

Comment (heavengate) [reply python/paddle/fluid/dataloader/batch_sampler.py]: Done, thanks!

File: paddle/fluid/pybind/imperative.cc [hunk 9]
```diff
@@ -225,11 +226,22 @@ void BindImperative(py::module *m_ptr) {
 
 #ifndef _WIN32
   // Dygraph DataLoader signal handler
-  m.def("_set_process_pid", [](int64_t key, pid_t pid) {
-    imperative::SetLoadProcessPID(key, pid);
+  m.def("_set_process_pids", [](int64_t key, py::object &obj) {
+    PADDLE_ENFORCE_EQ(
+        py::isinstance<py::tuple>(obj) || py::isinstance<py::list>(obj), true,
+        platform::errors::InvalidArgument(
+            "The batch data read into DataLoader is illegal."
```
Comment (chenwhql) [inline paddle/fluid/pybind/imperative.cc]: Here it is said that DataLoader child process info set type is illegal, `batch data read into DataLoader` is easily misunderstood

File: python/paddle/fluid/dataloader/batch_sampler.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                 data_source=None,
+                 indices=None,
+                 shuffle=False,
+                 batch_size=1,
+                 drop_last=False):
+        if data_source is None:
+            assert indices is not None, \
+                "either data_source or indices should be set"
+            assert isinstance(indices, list) or isinstance(indices, tuple), \
+                "indices should be a list or tuple"
```
Comment (chenwhql) [inline python/paddle/fluid/dataloader/batch_sampler.py]: add the ture indices type to error message?

File: python/paddle/fluid/dataloader/batch_sampler.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            self.indices = indices
+        else:
+            assert indices is None, \
+                "should not set both data_source and indices"
+            self.indices = list(range(len(data_source)))
```
Comment (chenwhql) [inline python/paddle/fluid/dataloader/batch_sampler.py]: Need to check the type of `data_source`?

File: python/paddle/fluid/dataloader/batch_sampler.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+        assert isinstance(batch_size, int) and batch_size > 0, \
+                "batch_size should be a positive integer"
+        self.batch_size = batch_size
+        assert isinstance(shuffle, bool), \
+                "shuffle should be a boolean value"
```
Comment (chenwhql) [inline python/paddle/fluid/dataloader/batch_sampler.py]: add receive type is better

File: python/paddle/fluid/dataloader/batch_sampler.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        self.shuffle = shuffle
+        assert isinstance(drop_last, bool), \
+                "drop_last should be a boolean number"
```
Comment (chenwhql) [inline python/paddle/fluid/dataloader/batch_sampler.py]: boolean number -> boolean value?

File: python/paddle/fluid/dataloader/dataset.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    by some implements of `fluid.io.BatchSampler`
+
+    see `fluid.io.DataLoader`.
+    """
+
+    def __getitem__(self, idx):
+        raise NotImplementedError("'{}' not implement in class "\
+                "{}".format('__getitem__', self.__class__.__name__))
+
+    def __len__(self):
```
Comment (chenwhql) [inline python/paddle/fluid/dataloader/dataset.py:71]: Is `IterDataset` implemented next time? PyTorch does not set `__len__` as the default function

File: python/paddle/fluid/dataloader/dataloader_iter.py [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+        # 2. single-process mode: read mini-batch data in main process
+        self._blocking_queue = None
+        self._thread = None
+        self._thread_done_event = threading.Event()
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        return len(self.batch_sampler)
```
Comment (guoshengCS) [inline python/paddle/fluid/dataloader/dataloader_iter.py]: Should it be `len(self._batch_sampler)`

File: python/paddle/fluid/dataloader/batch_sampler.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+    Batch sampler used by `fluid.io.DataLoader` should be a subclass
+    of `fluid.io.BatchSampler`, BatchSampler subclasses should
+    implement following methods:
+
+    :math:`__iter__`: return mini-batch indices iterably.
+    :math:`__len__`: get mini-batch number in an epoch.
+
+
+    Args:
+        data_source(Dataset): this could be a `fluid.io.Dataset`
```
Comment (LielinJiang) [inline python/paddle/fluid/dataloader/batch_sampler.py]: why not use `dataset` directly.

**Comment by phlrain** [comment]
这个优化有一个很重要的目标是为了提升在一些任务上的数据读取性能，最理想的目标是能够将 read_next的开销降为0，建议补充下，在一些经典的cv任务上read_next时间开销和竞品、优化前、以及最理想的的时间对比

File: python/paddle/fluid/dataloader/dataset.py (hunk 14 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/dataset.py:71]: For map-style dataset currently, `__len__` is required, I think it is better to set `__len__` not as default when add `IterDataset`

File: paddle/fluid/pybind/imperative.cc (hunk 9 above)
Comment (heavengate) [reply paddle/fluid/pybind/imperative.cc]: Done, thanks!

File: python/paddle/fluid/dataloader/batch_sampler.py (hunk 10 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/batch_sampler.py]: Done, thanks!

File: python/paddle/fluid/dataloader/batch_sampler.py (hunk 11 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/batch_sampler.py]: Done, thanks!

File: python/paddle/fluid/dataloader/batch_sampler.py (hunk 12 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/batch_sampler.py]: Done, thanks!

File: python/paddle/fluid/dataloader/batch_sampler.py (hunk 13 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/batch_sampler.py]: Done, thanks!

File: python/paddle/fluid/dataloader/dataloader_iter.py (hunk 15 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/dataloader_iter.py]: Done, thanks!

File: python/paddle/fluid/dataloader/batch_sampler.py (hunk 16 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/batch_sampler.py]: Done, thanks!

File: python/paddle/fluid/dataloader/datasets/mnist.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+        mode(str): 'train' or 'test' mode. Default 'train'.
+        download(bool): whether auto download mnist dataset if
+            :attr:`image_path`/:attr:`label_path` unset. Default
+            True
+
+    Examples:
+        
+        .. code-block:: python
+
+            from paddle.fluid.io import MNIST
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/datasets/mnist.py]: Need to discuss  where to  put new dataset

File: python/paddle/fluid/reader.py [hunk 18]
```diff
@@ -177,6 +98,247 @@ def __next__(self):
...
+
+    Args:  
+        dataset: the dataset to load data from, should be an instance
+            of subclass of `fluid.io.Dataset`.
+        feed_list (list(Variable)|tuple(Variable)): feed variable list.
+            The variables should be created by :code:`fluid.data()`.
+            This should be None in dygraph graph mode, but cannot be
+            None in static graph mode. Default None.
+        places(list(Place)|tuple(Place)): a list of Place, to put data
+            onto. Default None.
```
Comment (qingqing01) [inline python/paddle/fluid/reader.py]: From flowing assert, the places is must be set. Please tell user more clearly.

File: python/paddle/fluid/dataloader/dataloader_iter.py [hunk 19]
```diff
@@ -0,0 +1,537 @@
...
+    def _thread_loop(self):
+        try:
+            for indices in self._sampler_iter:
+                # read data from dataset in mini-batch
+                batch = [self._dataset[i] for i in indices]
+                if self._collate_fn is not None:
+                    batch = self._collate_fn(batch)
+
+                # batch each field
+                slots = []
```
Comment (guoshengCS) [inline python/paddle/fluid/dataloader/dataloader_iter.py]: 这里的处理是否需要调整下，`batch = self._collate_fn(batch)`后`batch`中每个field感觉应该已经是成batch的了，比如在`_collate_fn`中进行了padding。`batch each field`这段我理解是默认的`_collate_fn`，两者是否应该是ifelse的关系呢

File: python/paddle/fluid/dataloader/dataloader_iter.py [hunk 20]
```diff
@@ -0,0 +1,537 @@
...
+                        init_exception = None
+                    else:
+                        batch = [dataset[i] for i in indices]
+                        if self._collate_fn is not None:
+                            batch = self._collate_fn(batch)
+                except Exception as e:
+                    out_queue.put((idx, e))
+                else:
+                    # copy batch data to shared memory
+                    slots = []
```
Comment (guoshengCS) [inline python/paddle/fluid/dataloader/dataloader_iter.py]: 同上，`for items in batch`这段和`_collate_fn`两者是否应该是ifelse的关系呢，`for items in batch`作为默认的`_collate_fn`。

也参考了下这里 https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L226
```python
        if collate_fn is None:
            if self._auto_collation:
                collate_fn = _utils.collate.default_collate
            else:
                collate_fn = _utils.collate.default_convert

        self.collate_fn = collate_fn
```

File: python/paddle/fluid/reader.py [hunk 21]
```diff
@@ -177,6 +98,247 @@ def __next__(self):
...
+        return_list (bool): whether the return value on each device is 
+            presented as a list. It is only valid when iterable=True.
```
Comment (qingqing01) [inline python/paddle/fluid/reader.py]: There is no `iterable` in `__init__`

File: python/paddle/fluid/dataloader/dataloader_iter.py (hunk 19 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/dataloader_iter.py]: Done, thanks!

File: python/paddle/fluid/dataloader/dataloader_iter.py (hunk 20 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/dataloader_iter.py]: Done, thanks!

File: python/paddle/fluid/reader.py (hunk 18 above)
Comment (heavengate) [reply python/paddle/fluid/reader.py]: Done, thanks!

File: python/paddle/fluid/reader.py (hunk 21 above)
Comment (heavengate) [reply python/paddle/fluid/reader.py]: Done, thanks!

**Comment by heavengate** [comment]
> 这个优化有一个很重要的目标是为了提升在一些任务上的数据读取性能，最理想的目标是能够将 read_next的开销降为0，建议补充下，在一些经典的cv任务上read_next时间开销和竞品、优化前、以及最理想的的时间对比

已添加加速性能测试~

File: python/paddle/fluid/reader.py [hunk 22]
```diff
@@ -177,6 +98,264 @@ def __next__(self):
...
+
+    DataLoader supports single-process and multi-prcess data loading,
+    multi-process workers will be used to load data asynchronously if
+    :attr:`num_workers` is set as a positive number.
+
+    DataLoader only supports map-style dataset(can get a sample from
+    dataset with a given index) currently, for a map-style dataset,
+    please see :code:`fluid.io.Dataset`.
+
+    batch_sampler please se :code:`fluid.io.BatchSampler`
```
Comment (chenwhql) [inline python/paddle/fluid/reader.py]: `se` -> `see`?

Comment (heavengate) [reply python/paddle/fluid/reader.py]: Done, thanks!

**Review by chenwhql — DISMISSED** [review]
LGTM

**Review by Boyan-Liu — DISMISSED** [review]

File: python/paddle/fluid/dataloader/datasets/mnist.py (hunk 17 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/datasets/mnist.py]: Done, put new API under `paddle.io`

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by qingqing01 — APPROVED** [review]

**Review by Boyan-Liu — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**